sklearn/preprocessing/data.py

```diff
@@ -1901,3 +1906,288 @@ def transform(self, X):
+
+    def _sparse_fit(self, X):
+        """Compute percentiles for sparse matrices.
+
+        Parameters
+        ----------
+        X : sparse matrix, shape (n_samples, n_features)
+            The data used to scale along the features axis. The sparse matrix
+            needs to be semi-positive.
+        """
+        rng = check_random_state(self.random_state)
+
+        n_samples, n_feat = X.get_shape()
+        if self.subsample < n_samples:
+            subsample_idx = rng.choice(n_samples, self.subsample,
+                                       replace=False)
+            X_csr = X.tocsr()[subsample_idx]
+            X = X_csr.tocsc()
+
+        self.references_ = np.linspace(0, 1, self.n_quantiles,
+                                       endpoint=True)
+        # FIXME: it does not take into account the zero in the computation
+        self.quantiles_ = np.array([np.percentile(
+            X.data[X.indptr[feat]:X.indptr[feat + 1]], self.references_ * 100)
+                                    for feat in range(n_feat)]).T
+
+    def fit(self, X, y=None):
+        """Compute the quantiles used for normalizing.
+
+        Parameters
+        ----------
+        X : ndarray or sparse matrix, shape (n_samples, n_features)
+            The data used to scale along the features axis. If a sparse
+            matrix is provided, it will be converted into a sparse
+            ``csc_matrix``. Additionally, the sparse matrix needs to be
+            semi-positive.
+
+        Returns
+        -------
+        self : object
+            Returns self
+        """
+        X = self._validate_X(X)
+
+        if sparse.issparse(X):
+            self._sparse_fit(X)
+        else:
+            self._dense_fit(X)
+
+        self._build_f()
+
+        return self
+
+    def _dense_transform(self, X, direction=True):
+        """Forward and inverse transform for dense matrices.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            The data used to scale along the features axis.
+
+        direction : bool, optional (default=True)
+            If True, apply forward transform. If False, apply
+            inverse transform.
+
+        Returns
+        -------
+        Xt : ndarray, shape (n_samples, n_features)
+            Projected data.
+        """
+        Xt = X.copy()
+        if direction:
+            func_transform = self.f_transform_
+        else:
+            func_transform = self.f_inverse_transform_
+
+        for feat_idx, f in enumerate(func_transform):
+            Xt[:, feat_idx] = f(Xt[:, feat_idx])
+
+        return Xt
+
+    def _sparse_transform(self, X, direction=True):
+        """Forward and inverse transform for sparse matrices.
+
+        Parameters
+        ----------
+        X : sparse matrix, shape (n_samples, n_features)
+            The data used to scale along the features axis. The sparse matrix
+            needs to be semi-positive.
```

Comment:
You should make it explicit that it only works for CSC sparse matrices (I know this is not public API but it makes it easier to understand how the code works).